Allow `containerd` config imports

ndbaker1 opened the pull request:
**Issue #, if available:**
https://github.com/awslabs/amazon-eks-ami/issues/1628

**Description of changes:**

adding an import path for containerd to read for merging config tomls: https://github.com/containerd/containerd/blob/main/docs/man/containerd-config.toml.5.md#complete-configuration

avoid the need for users to directly edit the containerd config toml.

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

<!-- If this is a security issue, please do not discuss on GitHub. Please report any suspected or confirmed security issues to AWS Security https://aws.amazon.com/security/vulnerability-reporting/ -->

**Testing Done**

<!-- Include information regarding the testing that was completed with this changes. Where applicable, include details steps to replicate. -->

*[See this guide for recommended testing for PRs.](../doc/CONTRIBUTING.md#testing-changes) Some tests may not apply. Completing tests and providing additional validation steps are not required, but it is recommended and may reduce review time and time to merge.*

cartermckinnon commented:
/ci

cartermckinnon commented:
We need a user guide section about this, this should be the preferred way of amending the EKS containerd defaults.

dims commented:
@ndbaker1 do we create the directory? (is `/etc/containerd/config.d/` already present on disk)

ndbaker1 commented:
> We need a user guide section about this, this should be the preferred way of amending the EKS containerd defaults on AL2.

updated the user guide 👍 PTAL

> do we create the directory?

we weren't, just added this to `bootstrap.sh` to allow users to assume it exists when piping/writing to it

cartermckinnon commented on doc/USER_GUIDE.md:
```diff
@@ -296,6 +296,57 @@ $ curl -sSL "http://localhost:8001/api/v1/nodes/ip-192-168-92-220.us-east-2.comp
+If you have a containerd config of your own, you can place it inside the directory `/etc/containerd/config.d/` via your user data.
+Make sure that you explicitly set the `.toml` extension on the file.
+
+> [!NOTE]
+> When adding a new `containerd` config file to the `import` path after `containerd` has already started (i.e. `bootstrap.sh` already ran),
+> you'll need to restart `containerd` to pick up the latest configuration.
```
```suggestion
> [!NOTE]
> If you create an additional configuration file after `containerd` has already started (i.e. `bootstrap.sh` has already executed),
> you'll need to restart `containerd` to pick up the latest configuration.
```

cartermckinnon commented on doc/USER_GUIDE.md:
```diff
@@ -296,6 +296,57 @@ $ curl -sSL "http://localhost:8001/api/v1/nodes/ip-192-168-92-220.us-east-2.comp
 
+## Customizing Containerd Config
+
+In some cases, customers may want to customize the [containerd configuration](https://github.com/containerd/containerd/blob/main/docs/man/containerd-config.toml.5.md)
+on their nodes. We utilize the drop-in style `import` files [supported by containerd](https://github.com/containerd/containerd/blob/main/docs/man/containerd-config.toml.5.md#format)
+to achieve this. 
+
+**Import a containerd config**
+
+If you have a containerd config of your own, you can place it inside the directory `/etc/containerd/config.d/` via your user data.
+Make sure that you explicitly set the `.toml` extension on the file.
```
```suggestion
The EKS defaults for `containerd` will be written to `/etc/containerd/config.toml`. Additional configuration files placed in the `/etc/containerd/config.d/` directory will be imported by the default configuration. Additional configuration files will override the default as described in the [`containerd` documentation](https://github.com/containerd/containerd/blob/release/1.7/docs/man/containerd-config.toml.5.md).
```

cartermckinnon commented on doc/USER_GUIDE.md:
```diff
@@ -296,6 +296,57 @@ $ curl -sSL "http://localhost:8001/api/v1/nodes/ip-192-168-92-220.us-east-2.comp
+> When adding a new `containerd` config file to the `import` path after `containerd` has already started (i.e. `bootstrap.sh` already ran),
+> you'll need to restart `containerd` to pick up the latest configuration.
+
+The `bootstrap.sh` script will write the defaults for the containerd config to `/etc/containerd/config.toml`, containing an `import` 
+list containing the path glob: `/etc/containerd/config.d/*.toml`. `bootstrap.sh` will also create this directory,
+so the user is not required create it when executing commands after the bootstrap is run.
```
Don't think we need to explain all this here, but lets make sure that directory is created when the AMI is built, not by the bootstrap script.
```suggestion
```

ndbaker1 replied:
i don't want to omit any info about the pattern/file-ext, Ill just change the pattern to `/etc/containerd/config.d/*` and then ill make this change

cartermckinnon replied:
Sure, I don't feel too strongly about that detail, but I don't think anyone would expect files without a `.toml` extension to be included, all containerd config files are TOML files.

cartermckinnon commented on doc/USER_GUIDE.md:
````diff
@@ -296,6 +296,57 @@ $ curl -sSL "http://localhost:8001/api/v1/nodes/ip-192-168-92-220.us-east-2.comp
+can use the containerd cli:
+```
+$ containerd config dump
+
+version = 2
+root = "/var/lib/containerd"
+state = "/run/containerd"
+oom_score = 0
+imports = ["/etc/containerd/config.d/*.toml"]
+
+[grpc]
+  address = "/run/containerd/containerd.sock"
+  uid = 0
+  gid = 0
+
+[debug]
+  address = "/run/containerd/debug.sock"
+  uid = 0
+  gid = 0
+  level = "info"
````
```suggestion
```

cartermckinnon commented on doc/USER_GUIDE.md:
```diff
@@ -296,6 +296,57 @@ $ curl -sSL "http://localhost:8001/api/v1/nodes/ip-192-168-92-220.us-east-2.comp
+list containing the path glob: `/etc/containerd/config.d/*.toml`. `bootstrap.sh` will also create this directory,
+so the user is not required create it when executing commands after the bootstrap is run.
+
+> [!IMPORTANT]
+> containerd `imports` is the recommended approach to configure contianerd on EKS Optimized AMIS.  
+> Please attempt to use this method before ever making direct edits to `/etc/containerd/config.toml`.
```
```suggestion
> [!CAUTION]
> Making direct edits to the EKS default `containerd` configuration file is not recommended.
```

cartermckinnon reviewed (changes requested):
Think the wording can be tightened up and most of the details left in the code.

cartermckinnon commented on files/containerd-config.toml:
```diff
@@ -1,6 +1,7 @@
 version = 2
 root = "/var/lib/containerd"
 state = "/run/containerd"
+imports = ["/etc/containerd/config.d/*"]
```
I think we still want the `.toml` filter here, for sanity.

dims replied:
i agree, there has been a long debate on temporary files when some folks use utilities that create them some of the k8s config/yaml files

ndbaker1 replied:
gotcha, done!

cartermckinnon commented on files/bootstrap.sh:
```diff
@@ -555,6 +555,7 @@ if [[ "$CONTAINER_RUNTIME" = "containerd" ]]; then
   fi
 
   sudo mkdir -p /etc/containerd
+  sudo mkdir -p /etc/containerd/config.d
```
Do we need this if it's created at build time?

ndbaker1 replied:
bottom line, probably not, but its still a contract set with the bootstrap.sh and not the build so i figure its safe to have in both 🤔

cartermckinnon replied:
Fair enough!

cartermckinnon commented:
I changed the base branch to `main` and this is gonna need a rebase + some TLC. Sorry!

ndbaker1 commented:
thanks for the reminder! rebased now @cartermckinnon

phillebaba commented:
Could we get an eye on this :smiley: it would be nice to get a release with this fix so that we can resolve upstream issues.

Issacwww commented:
/ci

ndbaker1 commented:
/ci
